Sources/libhostmgr/S3Manager.swift

```diff
@@ -0,0 +1,175 @@
+                return []
+            }
+
+            return objects.compactMap { $0.toS3Object }
+        }
+    }
+
+    public func lookupObject(atPath path: String) async throws -> S3Object? {
+        try await withS3Client {
+            let request =  SotoS3.S3.HeadObjectRequest(bucket: bucket, key: path)
+            let result = try await $0.headObject(request)
+            return S3Object(key: path, size: Int(result.contentLength!))
+        }
+    }
+
+    public func download(
+        object: S3Object,
+        to destination: URL,
+        progressCallback: FileTransferProgressCallback?
+    ) async throws {
+
+        let signedURL = try await presignedUrl(forObject: object)
+
+        let destinationResolver: DownloadRequest.Destination = { _, _ in
+            return (FileManager.default.temporaryFilePath(), [.createIntermediateDirectories, .removePreviousFile])
+        }
+
+        let temporaryFile = try await AF
+            .download(signedURL, method: .get, to: destinationResolver)
+            .downloadProgress { progressCallback?(.progressData(from: $0)) }
+            .serializingDownloadedFileURL(automaticallyCancelling: true)
+            .value
+
+        _ = try FileManager.default.replaceItemAt(destination, withItemAt: temporaryFile)
+    }
+
+    public func download(object: S3Object) async throws -> Data? {
+        let signedURL = try await presignedUrl(forObject: object)
+        return await AF.request(signedURL, method: .get)
+            .validate()
+            .serializingData()
+            .response
+            .value
+    }
+
+    private func presignedUrl(forObject object: S3Object) async throws -> URL {
+        try await withS3Client {
+            var unsignedUrl = URL(string: "https://\(bucket).s3.\(region).amazonaws.com/\(object.key)")!
+
+            if try await bucketSupportsAcceleratedDownload {
+                unsignedUrl = URL(string: "https://\(bucket).s3-accelerate.amazonaws.com/\(object.key)")!
+            }
+
+            return try await $0.signURL(
+                url: unsignedUrl,
+                httpMethod: .GET,
+                expires: .hours(24)
+            )
+        }
+    }
+
+    private var bucketSupportsAcceleratedDownload: Bool {
+        get async throws {
+            try await withS3Client {
+                let request = SotoS3.S3.GetBucketAccelerateConfigurationRequest(bucket: self.bucket)
+                let response = try await $0.getBucketAccelerateConfiguration(request)
+                return response.status == .enabled
+            }
+        }
+    }
+
+    private func withS3Client<T>(_ block: (SotoS3.S3) async throws -> T) async throws -> T {
+        let awsClient = AWSClient(credentialProvider: credentialProvider, httpClientProvider: .createNew)
+        let s3Client = SotoS3.S3(client: awsClient, region: Region(rawValue: self.region))
+        let result = try await block(s3Client)
+        try await awsClient.shutdown()
```

Comment:
Having a `shutdown` API makes me think it's expensive to create a `S3` instance, not sure if that's the case and how expensive it is though... But do you think we should have a stored property holding onto one single `S3` instance, rather than creating many adhoc ones, if that makes a difference performance wise?

Reply:
This is mostly because it uses NIO under the hood (which is part of why we want to drop SotoS3!) 

The `shutdown` method kills any in-flight requests, and to do so it needs to wait for the event loop to finish processing. 

In _theory_ this should be totally free, because we only ever dispatch a single HTTP request then shut down the connection. Unfortunately we pay the API complexity cost though (for now).